File: backend/app/pipeline/chunker.py
import re

MAX_CHARS = 2800


def chunk_text(text: str) -> list[str]:
    sentences = re.split(r"(?<=[.!?])\s+", text.strip())
    chunks: list[str] = []
    current: list[str] = []

    for sentence in sentences:
        sentence = sentence.strip()
        if not sentence:
            continue

        trial = " ".join(current + [sentence]) if current else sentence
        if len(trial) > MAX_CHARS and current:
            chunks.append(" ".join(current))
            current = [sentence]
            continue

        if len(sentence) > MAX_CHARS:
            if current:
                chunks.append(" ".join(current))
                current = []
            for i in range(0, len(sentence), MAX_CHARS):
                chunks.append(sentence[i:i + MAX_CHARS])
        else:
            current.append(sentence)

    if current:
        chunks.append(" ".join(current))

    return chunks

File: app/src/main/java/com/ttsengine/pipeline/Chunker.kt
package com.ttsengine.pipeline

object Chunker {
    const val MAX_CHARS = 2800

    fun chunk(text: String): List<String> {
        val sentences = text.split(Regex("(?<=[.!?])\\s+"))
        val chunks = mutableListOf<String>()
        val current = StringBuilder()

        for (sentence in sentences) {
            val trimmed = sentence.trim()
            if (trimmed.isEmpty()) continue

            if (current.length + trimmed.length > MAX_CHARS && current.isNotEmpty()) {
                chunks.add(current.toString().trim())
                current.clear()
            }

            if (trimmed.length > MAX_CHARS) {
                trimmed.chunked(MAX_CHARS).forEach { chunks.add(it.trim()) }
            } else {
                if (current.isNotEmpty()) current.append(" ")
                current.append(trimmed)
            }
        }

        if (current.isNotEmpty()) {
            chunks.add(current.toString().trim())
        }

        return chunks
    }
}
